### PR TITLE
Rework into a more flexible listener

### PR DESCRIPTION
**Pull request by jefferai**

### CODE REVIEW

File: net/splitlistener.go [hunk 1]
```diff
@@ -73,81 +70,177 @@ func NewSplitListener(baseLn net.Listener) *SplitListener {
...
-// NodeEnrollmentListener returns the listener receiving connections related to
-// this library
-func (l *SplitListener) NodeEnrollmentListener() net.Listener {
-	return l.nodeeBabyListener
+		// Not authenticated
+		val, ok := l.babyListeners.Load(UnauthenticatedNextProto)
+		if !ok {
+			_ = conn.Close()
+		} else {
+			val.(*babySplitListener).incoming <- splitConn{conn: conn}
```
[the remark this answers is not in the document]
Comment (jefferai) [reply net/splitlistener.go:188]: Good catch, will do

File: net/splitlistener.go [hunk 2]
```diff
@@ -11,59 +11,56 @@ import (
...
-// It is required that the base listener return *tls.Conn values on Accept.
-//
-// This package can be useful for integration with systems that expect to do
+// SplitListener can be useful for integration with systems that expect to do
 // their own handling of connections off of a net.Listener. One such example is
 // gRPC which expects to be handed a listener and has deprecated any ability to
-// simply hand it a connection. The NodeEnrollmentListener can be given to the
-// gRPC server and the OtherListener can be used for other purposes.
+// simply hand it a connection. GetListener can be called with
+// AuthenticatedNonSpecificNextProto can be given to the gRPC server to pass
```
Comment (jimlambrt) [inline net/splitlistener.go:30]:
```suggestion
// AuthenticatedNonSpecificNextProto which in turn can be given to the gRPC server to pass
```

File: net/splitlistener.go [hunk 3]
```diff
@@ -11,59 +11,56 @@ import (
...
+// authenticated connections to gRPC, and a listener with
+// UnauthenticatedNextProto can be passed to another handler.
+//
+// SplitListener is compatible with the protocol package's Dialer's
+// WithExtraAlpnProtos option. If the base listener is a protocol.Listener, the
+// client-specified NextProtos will be passed through to here and used to allow
+// further switching based on listeners retrieved from GetListener with custom
+// protos.
+//
+// Regardless of client-specified NextProto or not, any connections that return
```
Comment (jimlambrt) [inline net/splitlistener.go:40]:
```suggestion
// Regardless of client-specified NextProto or not, any connections that's return
```

File: net/splitlistener.go [hunk 4]
```diff
@@ -73,81 +70,177 @@ func NewSplitListener(baseLn net.Listener) *SplitListener {
...
 			continue
 		}
 
 		if !tlsConn.ConnectionState().HandshakeComplete {
 			if err := tlsConn.Handshake(); err != nil {
 				// Without a successful handshake we can't know which proto;
-				// send down the other listener
-				_ = tlsConn.Close()
-				l.otherBabyListener.incoming <- splitConn{err: temperror.New(fmt.Errorf("tls handshake resulted in error: %w", err))}
+				// send down the unauth listener
```
Comment (jimlambrt) [inline net/splitlistener.go:117]:
```suggestion
				// send it down to the unauth listener
```

File: net/splitlistener.go [hunk 5]
```diff
@@ -73,81 +70,177 @@ func NewSplitListener(baseLn net.Listener) *SplitListener {
...
+		} else {
+			val.(*babySplitListener).incoming <- splitConn{conn: conn}
+		}
+	}
 }
 
-// OtherListener returns the listener receving all other connections
-func (l *SplitListener) OtherListener() net.Listener {
-	return l.otherBabyListener
+// GetListener returns a listener that will send connections that satisfy both
```
Comment (jimlambrt) [inline net/splitlistener.go:193]:
```suggestion
// GetListener returns a listener that will return connections that satisfy both
```

**Review by jimlambrt — COMMENTED** [review]
I few minor comments.  overall, this was actually not too difficult to understand.  

I wish it had better test coverage. especially in `Start(..)`

**Pull request merged**